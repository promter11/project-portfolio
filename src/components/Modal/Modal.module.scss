@import "../../styles/_variables";
@import "../../styles/_mixins";

.modal {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 5;
  @include flex(nowrap, row, center, center);

  &.modalEnter,
  &.modalExitActive,
  &.modalExitDone {
    .overlay {
      opacity: 0;
      transition: opacity 300ms;
    }

    .content {
      opacity: 0;
      transform: scale(0);
      transition: opacity 300ms, transform 300ms;
    }
  }

  &.modalEnterActive,
  &.modalEnterDone,
  &.modalExit {
    .overlay {
      opacity: 1;
      transition: opacity 300ms;
    }

    .content {
      opacity: 1;
      transform: scale(1);
      transition: opacity 300ms, transform 300ms;
    }
  }

  .overlay {
    position: fixed;
    z-index: -5;
    @include width-height(100vw, 100vh);
    background-color: rgba(0, 0, 0, 0.5);
  }

  .content {
    position: relative;
    max-width: 77rem;
    width: calc(100% - 3rem);

    .button {
      position: absolute;
      top: 2rem;
      right: 2rem;
      background-color: transparent;

      .icon {
        @include width-height(2rem, 2rem);
      }
    }

    .modalWrapper {
      display: flex;

      .modalBlock {
        width: 50%;

        &:last-child {
          padding: 4rem;
          border-radius: 0 0.8rem 0.8rem 0;
          background-color: $white-color;
        }

        .modalImage {
          @include width-height(100%, 100%);
          border-radius: 0.8rem 0 0 0.8rem;
          object-fit: cover;
        }

        .modalTitle {
          margin-bottom: 2rem;
          text-transform: uppercase;
        }

        .modalDesc {
          margin-bottom: 2rem;
          @include text(1.4rem, 100, 150%, none);
        }

        .modalList {
          list-style-type: disc;
          margin-bottom: 2rem;
          padding-left: 1.5rem;

          .modalListItem {
            @include text(1.4rem, normal, 150%, none);
          }
        }

        .modalLink {
          @include button(7.6rem, -7.6rem, $white-color, $black-color);
        }
      }
    }
  }
}
