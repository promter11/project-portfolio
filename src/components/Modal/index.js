import React from "react";
import { bool, func, node } from "prop-types";
import { createPortal } from "react-dom";
import { CSSTransition } from "react-transition-group";

import Close from "../../assets/icons/portfolio/close.svg";
import styles from "./Modal.module.scss";

const modalRoot = document.getElementById("modal-root");

const Modal = ({ children, modalStatus, toggleModal }) => {
  return createPortal(
    <CSSTransition
      in={modalStatus}
      timeout={500}
      classNames={{
        enter: styles.modalEnter,
        enterActive: styles.modalEnterActive,
        exitActive: styles.modalExitActive,
      }}
      unmountOnExit
    >
      <div className={styles.modal}>
        <div className={styles.overlay} onClick={toggleModal} />
        <div className={styles.content}>
          <button className={styles.button} type="button" onClick={toggleModal}>
            <img className={styles.icon} src={Close} alt="Close" />
          </button>
          {children}
        </div>
      </div>
    </CSSTransition>,
    modalRoot
  );
};

Modal.propTypes = {
  children: node,
  modalStatus: bool,
  toggleModal: func,
};

export default Modal;
