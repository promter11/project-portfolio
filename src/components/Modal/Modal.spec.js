import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";

import Modal from "./index";

const props = {
  modalStatus: false,
  toggleModal: jest.fn(),
};

describe("Modal component", () => {
  let wrapper;

  beforeEach(() => {
    ReactDOM.createPortal = jest.fn((element) => element);

    wrapper = shallow(<Modal {...props}>Content</Modal>);
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  it("renders correctly with props", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("simulates onClick event on toggleModal function", () => {
    wrapper.find(".overlay").simulate("click");

    expect(props.toggleModal).toHaveBeenCalled();
  });
});
