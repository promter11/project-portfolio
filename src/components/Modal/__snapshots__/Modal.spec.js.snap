// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders correctly with props 1`] = `
<CSSTransition
  classNames={
    Object {
      "enter": "modalEnter",
      "enterActive": "modalEnterActive",
      "exitActive": "modalExitActive",
    }
  }
  in={false}
  timeout={500}
  unmountOnExit={true}
>
  <div
    className="modal"
  >
    <div
      className="overlay"
      onClick={[MockFunction]}
    />
    <div
      className="content"
    >
      <button
        className="button"
        onClick={[MockFunction]}
        type="button"
      >
        <img
          alt="Close"
          className="icon"
          src="close.svg"
        />
      </button>
      Content
    </div>
  </div>
</CSSTransition>
`;
